libz-rs-sys/src/gz.rs

```diff
@@ -0,0 +1,230 @@
+                b'T' => state.direct = true,
+                _ => {} // for compatibility with zlib-ng, ignore unexpected characters in the mode
+            }
+        }
+    }
+
+    // Must specify read, write, or append
+    if state.mode == GzMode::GZ_NONE {
+        gz_free(state_ptr);
+        return ptr::null_mut();
+    }
+
+    // Can't force transparent read
+    if state.mode == GzMode::GZ_READ {
+        if state.direct {
+            gz_free(state_ptr);
+            return ptr::null_mut();
+        }
+        state.direct = true;
+    }
+
+    // Save the path name for error messages
+    // FIXME: support Windows wide characters for compatibility with zlib-ng
+    state.path = libc::strdup(path);
+
+    // Open the file unless the caller passed a file descriptor.
+    if fd > -1 {
+        state.fd = fd;
+    } else {
+        let mut oflag: c_int = O_LARGEFILE;
+        if cloexec {
+            oflag |= O_CLOEXEC;
+        }
+        if state.mode == GzMode::GZ_READ {
+            oflag |= O_RDONLY;
+        } else {
+            oflag |= O_WRONLY | O_CREAT;
+            if exclusive {
+                oflag |= O_EXCL;
+            }
+            if state.mode == GzMode::GZ_WRITE {
+                oflag |= O_TRUNC;
+            } else {
+                oflag |= O_APPEND;
+            }
+        }
+        // FIXME: support _wopen for WIN32
+        state.fd = libc::open(state.path, oflag, 0o666);
+        if state.fd == -1 {
+            gz_free(state.path);
+            gz_free(state_ptr);
+            return ptr::null_mut();
+        }
+        if state.mode == GzMode::GZ_APPEND {
+            libc::lseek(state.fd, 0, SEEK_END);
+            state.mode = GzMode::GZ_WRITE;
+        }
+    }
+
+    if state.mode == GzMode::GZ_READ {
+        // Save the current position for rewinding
+        state.start = libc::lseek(state.fd, 0, SEEK_CUR);
+        if state.start == -1 {
+            state.start = 0;
+        }
+    }
+
+    // FIXME verify the file headers, and initialize the inflate/deflate state
+
+    state_ptr.cast::<gzFile_s>()
+}
+
+/// Close an open gzip file and free the internal data structures referenced by the file handle.
+///
+/// # Safety
+/// This function may be called at most once for any file handle.
+#[cfg_attr(feature = "export-symbols", export_name = crate::prefix!(gzclose))]
+pub unsafe extern "C-unwind" fn gzclose(file: gzFile) -> c_int{
+    if file.is_null() {
+        return Z_STREAM_ERROR;
+    }
+
+    // FIXME: once read/write support is added, clean up internal buffers
+    let state = &*file.cast::<gz_state>();
+    let err = libc::close(state.fd);
+    gz_free(state.path);
+    gz_free(file);
+    if err == 0 {
+        Z_OK
+    } else {
+        Z_ERRNO
+    }
+}
+
+// Allocate a block of memory
+// # Safety: if the returned value is non-NULL, caller must delete it only with gz_free.
+unsafe fn gz_alloc<T>() -> *mut T {
+    libc::calloc(1, size_of::<T>()).cast::<T>()
```

Comment:
The caller must also not ask for any `T` whose alignment is larger than what malloc provides (I believe it always aligns to at least 8 (16 on most platforms), so the code here is fine in practice, but it is part of the safety contract.

I'd like to statically assert that fact, but with our minimum supported rust version we can't